### DISCUSSION TITLE
How can we improve the performance of the Cart & Checkout blocks?

### DISCUSSION THREAD

**Discussion by alexflorisca** [woocommerce/woocommerce#45034] [WooCommerce, Blocks, and Site Editing]
This is an open ended question to gather some feedback from the community about any performance issues that you may be facing. We're aware there is work to be done in this area and we want to improve. Please describe your concerns with this in mind and in as much detail as possible to help us dig into and address any issues that come up. 

We're specifically interested in any performance related issues that may be stopping merchants or developers from adopting the Cart & Checkout blocks over the shortcode version.

**Comment by lkraav** [comment]
Due to a Blocks integration bug in EU VAT Numbers plugin, I got a glimpse of what going to Blocks based checkout adds to JS asset payload.

![image](https://user-images.githubusercontent.com/147228/212064444-7c951827-e29c-4b07-969a-e9a956430451.png) :scream_cat: 

It seems you would be forced to have a full blown React app living on the frontend, whether your WP theme is using React in general or not.

I think this adds at least ~300 kB (compressed) JS payload (initial numbers, my measurement process is still ongoing).

We don't plan to convert our classic theme to a block theme any time soon, but still, I feel uneasy about this direction.

**Comment by alexflorisca** [comment]
Thanks for pointing that out @lkraav. 

I've done a bit of experimenting by loading a fresh WooCommerce Store, with the Twenty Twenty Three theme installed, an no other plugins. Here's some numbers for the Checkout page:

|                    	| Blocks Checkout 	| Shortcode Checkout 	| Shopify 	|
|--------------------	|-----------------	|--------------------	|---------	|
| Total Payload      	| 2.9MB           	| 935kb              	| ~6.1MB~ 4.1MB   	|
| Total Transferred  	| 2.1MB           	| 1.3MB*             	| ~3MB~ 1.9MB     	|
| Number of requests 	| 144             	| 77                 	| ~146~ 78     	|

The number of requests has almost doubled for Blocks, which isn't great so this is something that we can look into. I suspect the reason is because we rely on a few layers of abstraction on top - WooCommerce and Wordpress, each with their packages and set ways of doing certain things. We can investigate if we can simply this.

The total payload is almost 3 times bigger for Blocks than for Shortcode Checkout. That's a lot of JS - and you are right, it is a react app. The improvements in UI require a fair chunk of JS and we chose react as it's the basis of Gutenberg. It's worth noting here that blocks don't have to (and often don't) render react on the frontend. We do it here because of the complexity of the block. We can look at trimming this down and seeing if there are ways to bring the bundle size down.

What's surprising is that the Total Transferred payload is bigger than the total payload for the shortcode version (maybe some things are included twice? Yet to do more investigation there). So the actual transferred payload only increases by 800kb. 

I've also included the Shopify Checkout for comparison, and we can see we're doing slightly better which is re-assuring, given their success and work done on improving conversion rates.

**Reply by antonyjsmith** [reply]
I'm not sure where that figure for Shopify comes from but it's not even close to the actual checkout size I see on a fresh setup using the default Dawn theme. I know it's not hugely relevant to the discussion but the figures I see are much closers to the blocks checkout payload so circa 3MB but with around 70-80 requests. That's on a production store too which inculdes Stripe and Google Pay which are large payloads. By the sound of this assessment they are not included in the WooCommerce figures here.

**Reply by alexflorisca** [reply]
You're right @antonyjsmith I was testing on a friend's Shopify store which had different checkout options set up, inc Google Pay, Shop Pay, etc. I tested on a fresh store and updated the figures in the table. The figures for the shortcode and Blocks are both on a clean store with no other plugins installed. So the size isn't hugely different to Shopify in this case, but we could do with reducing the number of requests for sure.

**Comment by himanshu-mahendra-jain** [comment]
Yeah woo native blocks performance is abysmal. Not only in terms of loading the assets but REST api performance is also not upto the mark. May be caching is required here. I'll suggest you guys to check out how a 3rd party builder have implemented custom woo blocks based on Store API. The performance is much better and the number of assets loaded is pretty low. If they can do it then I am pretty sure core can also improve it.

CSS Requests on Cart Page:
![image](https://user-images.githubusercontent.com/101234934/213626827-f0e68069-7921-41da-81de-2a321a66d7a6.png)

JS Requests on Cart Page:
![image](https://user-images.githubusercontent.com/101234934/213626996-cfafdb86-b74a-4e9b-b2be-34f6a4cce67a.png)

jQuery-migrate and emoji request can be eliminated from these requests without any issue.

Builder: https://www.youtube.com/watch?v=iK1JLLxjC4w
Docs: https://docs.cwicly.com/woocommerce/cart
Sandbox to test: https://app.instawp.io/launch?t=cwicly

**Reply by alexflorisca** [reply]
Thanks for sharing your thoughts. Care to elaborate a bit on:

> Yeah woo native blocks performance is abysmal. Not only in terms of loading the assets but REST api performance is also not up to the mark.

Abysmal in terms of what? What could we do to improve it? We've talked about number of requests in the comment above and we will address that. Anything else you'd like to see? In your opinion what is cwicly doing we are not?

**Reply by himanshu-mahendra-jain** [reply]
Thanks for your reply @alexflorisca

Yes. I know you guys will be reducing the number of request which should reduce the overall payload size. I had posted Cwicly's example because they hardly load any asset compared to a standard woo blocks. Difference is massive - 200 KB vs 2 MB worth compressed assets loaded for the cart page. My suggestion is to review each asset loaded and load what is only required for the cart page for example.

**Comment by senadir** [comment]
I recently run a quick bundle analysis of a production page running Checkout block, exported the result and came to the following numbers (everything here is gzipped, as in, the size of what's sent from the server to the client):

The total JS loaded by us was 854 KB gzipped, it's broken down into:

Gutenberg packages: 547 KB gzipped.

GB packages included:

wp.components: 195.09 KB **(shouldn’t be there)**
wp.block-editor: 173.84 KB **(shouldn’t be there)**
wp.blocks: 50.1 KB **(shouldn’t be there)**
wp.vendors: 41.34 KB
wp.date: 32.26 KB **(shouldn’t be there)**
wp.compose: 12.45 KB
wp.rich-text: 10.85 KB
wp.data: 7.73 KB
wp.element: 4.98 KB
wp.dom: 4.75 KB
wp.vendors: 4.04 KB
wp.i18n: 3.82 KB
wp.url: 3.58 KB
wp.redux-routine: 2.82 KB
wp.vendors: 2.51 KB
wp.api-fetch: 2.29 KB
wp.autop: 2.22 KB
wp.plugins: 1.97 KB

Excluding what shouldn’t be there, we get **96 KB** which is kind of reasonable but still too much because it doesn’t include Checkout actual code.

For WordPress vendor packages, the total size was **97 KB**:
react-dom: 41.34 KB
react: 4.05 KB
inert-polyfill: 2.5 KB **(shouldn’t be there)**
lodash: 25.05 KB **(shouldn’t be there)**
moment: 19.1 KB **(shouldn’t be there)**
wp-polyfill: 6.58 KB
regenerator-runtime: 2.5 KB

WooCommerce Blocks actual size is **210 KB** gzipped, of which **27 KB** is a script that shouldn’t load.

All of the packages marked as shouldn't be there are going to be removed, we will also run more tree shaking on the actual WooCommerce Blocks packages to reduce size, we anticipate that the **854 KB will drop to 334 KB**, a **60%** reduction of the initial size.

**Reply by himanshu-mahendra-jain** [reply]
Is it possible to exclude a block from caching? I want to cache the whole cart and checkout page blocks except the cart and checkout block. That will also reduce the loading time. Something like ESI caching.

**Reply by straightvisions-matthias-bathke** [reply]
Due to the fact that the current behavior is worst practice (backend scripts loaded in frontend wtf), we should move this out of the brainstorming area into "fix as soon as possible" area. https://github.com/woocommerce/woocommerce/issues/37566

**Reply by alexflorisca** [reply]
Thanks for pointing this out @straightvisions-matthias-bathke. This has been fixed in #9251.

**Comment by jongore** [comment]
You can improve the performance of Cart and Checkout blocks by using the following steps.

1# Caching: Implementing caching can significantly reduce the number of requests made to the server, which can improve the performance of both Cart and Checkout blocks.

2# Lazy loading: Lazy loading involves only loading content when it is needed. For example, only loading product images when they are in the user's viewport. This can improve the initial load time of the page and reduce the number of requests.

3# Minimizing scripts and styles: Minimizing scripts and styles can help reduce the size of the page, which can improve load times. It's also important to ensure that scripts and styles are loaded asynchronously to avoid blocking the page.

4# Use a Content Delivery Network (CDN): Using a CDN can improve the performance of the Cart and Checkout blocks by reducing the distance between the user and the server, and by distributing the load across multiple servers.

**Reply by alexflorisca** [reply]
Thanks for your comments here @jongore. We do implement all of the above in some format. The browser caches assets on inital page load and we also prefetch some assets, for example when a user is on the Cart page, we make an assumption that the user will navigate to Checkout next and we prefetch the assets for the Checkout block. Server side and CDN caching is at the level of infrastructure which we don't control here. so that will be up to each Wordpress installation to manage.

Lazy loading isn't really applicable for the Cart & Checkout blocks as there are no heavy images or other assets on the page that would require this.

All our scripts are minimized in the production version of the plugin.

And again the CDN is difficult to control here as it depends on the where Wordpress itself is installed. 

Appreciate your input, any more suggestions are welcome

**Comment by jongore** [comment]
I work on Shopify and I use following points to improve the performance of Cart & Checkout blocks in Shopify

1- Optimize website images
2- Reduce installed apps
3- Minimize use of external scripts
4- Use faster theme
5- Enable AMP
6- Consider using CDN
7- Monitor website performance.

**Comment by straightvisions-matthias-bathke** [comment]
@jongore Thank you for your input. Your recommendations regarding performance optimisation are certainly a good orientation when one wants to improve a website overall. There are countless topics regarding PageSpeed that a webmaster should take into account, I have created a [PageSpeed FAQ](https://straightvisions.com/en/news/pagespeed-the-ultimate-faq/) for this as a guide.

Unfortunately, your recommendations do not solve the [actual problem](https://github.com/woocommerce/woocommerce/issues/37566) that unused assets (CSS/JS) are output by these Cart & Checkout blocks of WooCommerce. Thus, the cart & checkout blocks themselves violate best practices and need to be directly corrected by a patch and similar errors need to be avoided by better coding guidelines in the future.

There is no justification for loading the block editor as a dependency in the frontend. It should not be the task of thousands of webmasters to alleviate the symptoms of serious PageSpeed violations by WooCommerce. If a developer of a product used by thousands of websites worldwide can avoid unnecessarily loading hundreds of kilobytes of assets by a little more careful handling of script dependencies and on demand loading, then they should do so.

**Reply by alexflorisca** [reply]
Thanks for pointing out a serious issue on our side @straightvisions-matthias-bathke. There were some technical limitations with regards to dependencies earlier in the development of these blocks and performance wasn't taken into consideration as we aimed at building the core functionality. We are now trying to correct some of the mistakes and improve performance on every front so really appreciate your input here and taking the time to raise the issue [here](https://github.com/woocommerce/woocommerce-blocks/issues/9040) As pointed out by @lkraav below, this has now been fixed in #9251 

Please keep raising your concerns and suggestions here with regards to performance and unnecessary bloat, we're very keen to address these issues. You mentioned some unused JS as well? What are you referring to?

**Comment by ResponsiveFlow** [comment]
Has this issue been solved in the latest release?

**Reply by alexflorisca** [reply]
This has been partly fixed in #9251 which has been shipped in v10.3.0

**Comment by asafm7** [comment]
This might be related. Also reported here: https://github.com/woocommerce/woocommerce-blocks/issues/9711#issue-1740315276

When adding WooCommerce Filter by Attributes block to a non-WooCommerce page, these 4 files are being enqueued:

![image](https://github.com/woocommerce/woocommerce-blocks/assets/14274769/926aa086-7fe8-4b0e-b2e5-35cddccb0046)

Dequeuing either one of these files breaks the Filter by Attributes block, although "blocks-checkout", for example, isn't semantically related.

Maybe they are bound as dependencies.

Loading 0.5 MB of JS just to have a simple filter functionality (which isn't even an Ajax filter) feels redundant.

Maybe there is a better way to achieve it?

**Reply by alexflorisca** [reply]
Hi @asafm7 thanks for raising this. This is the same issue @straightvisions-matthias-bathke pointed out above. We had some dependencies on some external gutenberg components, but we were only using a few styles from those. This has been fixed in #9251 and the first 2 assets in your screenshot, `components.min.js` and `block-editor.min.js` should not be there anymore. The `blocks.min.js` and the `blocks-checkout.min.js` are still necessary, but we will review the code to see if we can optimise or split it better to avoid redundancy.

**Reply by asafm7** [reply]
Thanks, @alexflorisca. 

I'm wondering about these:
![image](https://github.com/woocommerce/woocommerce-blocks/assets/14274769/736cc63e-26b6-4fc9-94ba-410c148e9f41)

But I haven't researched. They might be essential.

Also, removing `cart` and `checkout` as dependencies could be great in some use cases, as WooCommerce is sometimes used as a simple catalog.

**Comment by asafm7** [comment]
I raised the question in core-performance Slack channel, and got this response:

> It looks like you're checking the dashboard/editor??
> These files should not be loaded on the frontend of your site... Unless a plugin/theme you're using is doing something weird

Maybe WooCommerce is doing it wrong?

https://wordpress.slack.com/archives/C02KGN5K076/p1686040160612489?thread_ts=1685744975.545929&cid=C02KGN5K076

**Reply by lkraav** [reply]
Possibly fixed in https://github.com/woocommerce/woocommerce-blocks/pull/9251

**Reply by asafm7** [reply]
Thanks @lkraav !

Looks promising. It seems to be about CSS though, will it also prevent loading of redundant JS?

**Reply by asafm7** [reply]
@lkraav I see the fixed version is available. I've checked, and it does solve the issue. Thanks!

**Reply by alexflorisca** [reply]
Yep, this has been fixed as of Blocks v10.3.0. There as also been another optimisation done by @Aljullu to only load block specific styles rather than one big stylesheet. This has shipped in v10.6.0.

> will it also prevent loading of redundant JS?

@asafm7 ~~what redundant JS are you referring to here?~~ I've just seen your comment above and replied inline

**Reply by Aljullu** [reply]
> This has shipped in v10.6.0.

Just a quick FYI that this change was finally bumped to v10.7.0 to provide some more time for testing. We are planning to publish a post on developer.woocommerce.com about it to raise awareness in the community. :slightly_smiling_face:

**Comment by savagecatfood** [comment]
Are we able to change the text size in the Woocommerce Cart block? The font size where it shows the shipping cost is too small.

**Reply by alexflorisca** [reply]
Hey @savagecatfood, there is no control for the individual rows of text in the order summary on the Cart block, but you can change the text size for the whole theme (assuming you are using a block theme) from the site editor like so:

https://github.com/woocommerce/woocommerce/assets/3966773/2fbaec36-8d79-4830-b59f-84a7fe24d180

Also this question isn't relevant to the performance discussion above, so in future just letting you know such questions can best be answered on WooCommerce slack in the #woocommerce-blocks-and-block-themes channel.
